README.md
# scheduler - caldav based timer on Raspbeey Pi
The idea of this project is to control different power switches via web-calender entries.
I'm using Nextcloud to host the calendar, but any other server with a caldav interfce should work.

The script needs to be called regularly by a cron job to check for calender entries and
schedule the switching events. The time interval for calling teh script is defined in
the configuration file /etc/caltimer/caltimer.ini

Supported switched:
* RC power plugs via 433 MHz transmitter (connected to Raspberry Pi GPIO)
* Switch conencted directly to a GPIO port
* Pulsed switch connected directly to the GPIO port
  * Can be used to emulate the power button of another device (I use it for my NAS)

## Config file
The Python script uses an ini file to define the available swithce and some general settings.
The ini file needs to be stored as /etc/caltimer/caltimer.ini

## Cron
If the interval is set to 15 minutes, a cron job needs to run every 15 min as well. Best is to start the
cron job about 1 min before each interval:
```
# calendar timer
14-59/15 *  * * *   root    /opt/caltimer/caltimer.py
```

## Timer entries
For each switch time a calendar entry is added with an aritrary summary, the location containing teh switch name
(as per ini file) and an optional description with extra settings.
The start and end time of the calendar entry are simply the on and off time for the switch.

## Dependencies
Some extra Python libraries used are.
* caldav https://pypi.python.org/pypi/caldav
  * https://github.com/python-caldav/caldav
  * Install with pip3 caldav
* sunrise_sunset library
  * https://github.com/jebeaudet/SunriseSunsetCalculator
  * sudo pip3 install git+https://github.com/palto42/SunriseSunsetCalculator.git

On Raspbian Jessie I installed pip3 for Python3, which required some extra libraries first:
* Install
  * sudo apt-get install libxml2-dev libxslt1-dev
* Install pip3
  * sudo apt-get install python3-pip

## RC transmitter
There are several web pages decribing how to use a cheap RF433 transmitter on Raspberry Pi.
I'm using the library https://github.com/ninjablocks/433Utils since this was the only one which worked with my ZAP switches.
Unfortunately this library onyl provides a binary an not python code, but the main thing is that it works well ;-)
The RF433 transmitter is conencted to GPIO 17 (WiringPi 0) and the receiver on GPIO 27 (WiringPi 2).

The receiver is only used to sniff the switch code if it is not known. It's not required for this scheduler script.

## libraries

### RF 433
The script supports the (rpi_rf)[https://github.com/milaq/rpi-rf] and the (433utils)[https://github.com/ninjablocks/433Utils] libraries to send at 433 MHz.
The (rpi_rf)[https://github.com/milaq/rpi-rf] provides a python library to send codes at 433 MHz, but it requries at least a Raspberry Pi 2B or newer (faster) for stable timing.
From the (433utils)[https://github.com/ninjablocks/433Utils] the "codesend" command line tool is used.


changelog.txt
# 2018-10-05
* Better combine argparse and configparser,
  see http://blog.vwelch.com/2011/04/combining-configparser-and-argparse.html
* Code clean-up


nanocul_rec_test.py
#!/usr/bin/python3
import serial
from time import sleep

#+ 01234567890123456789
#~ kr07C2AD1A30CC0F0328
#~ ||  ||||  ||    ++-------- Transmitter Code 2
#~ ||  ||||  ++-------------- Keycode
#~ ||  ++++------------------ Transmitter Code 1
#~ ++------------------------ kr wird von der culfw bei Empfang einer Kopp Botschaft als 

#port = "/dev/ttyUSB.FTDI"
#port = "/dev/ttyUSB.Nano"
port = "/dev/ttyAMA0"
ser = serial.Serial(port, 38400, timeout=0)

while True:
    data = ser.read(9999)
    if len(data) > 0:
        #print ('length:',len(data))
        #print ('Got:', data.decode('utf-8')[0:-2]) # decode ascii or utf-8, strip last 2 char (cr lf)
        if len(data) == 22:
            transmitter = data.decode('utf-8')[4:8] + data.decode('utf-8')[16:18]
            keycode = data.decode('utf-8')[10:12]
            print ('Transmitter:',transmitter,' Key:',keycode)
        else:
            print ('Got:', data.decode('utf-8')[0:-2]) # decode ascii or utf-8, strip last 2 char (cr lf)

    sleep(0.2)
    #print ('not blocked')

ser.close()


scheduler.py
#!/usr/bin/python3

#########################################################  
# Calernder based scheduler to switch RF433 sockets     #
# using cli codesend                                    #
#                                                       #
# Extra options can be defined in the description       #
# filed of each event, using an ini-style format:       #
#   [random]                                            #
#   # add random offset for start/end time              #
#   # values can be positive or negative                #
#   # "all" option is applied to start and end          # 
#   start : max random time in minutes                  #
#   end : max random time in minutes                    #
#   all : max random time in minutes                    #
#                                                       #
#   [sun]                                               #
#   # replace the start/end time with the current       #
#   # sunrise or sunset time +/- offset                 #
#   start : rise or set                                 #
#   end : rise or set                                   #
#   start_offset : offset in minutes                    #
#   end_offset : offset in minutes                      #
#                                                       #
# Matthias Homann                                       #
# 2017-10-03                                            #
#########################################################

import configparser
import subprocess
from random import uniform
import sched, time
from datetime import datetime, date, timedelta
from sunrise_sunset import SunriseSunset
import caldav
from caldav.elements import dav, cdav
import RPi.GPIO as GPIO
import logging, sys

# set logging destination and level
# level: CRITICAL > ERROR > WARNING INFO > DEBUG > NOTSET
# filename='example.log' or stream=sys.stderr
logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)
#logging.debug('A debug message!')
#logging.info('We processed %d records', 10)


def print_time(a='Time'):
    print(a, time.strftime('%H:%M:%S'))

def rc_switch(switch,onoff,stime):
    if onoff:
      sendcode=switches[switch]['oncode']
    else:
      sendcode=switches[switch]['offcode']
#    print('RC:',stime, sendcode)
    s.enterabs(stime,1,subprocess.call,
        argument=([switches['DEFAULT']['rf433'],sendcode,
        switches[switch]['protocol'],switches[switch]['pulselength']],));

def gpio_switch(switch,onoff,stime):
# Set the pin to output (just to be sure...)
    try:
      GPIO.setup(int(switches[switch]['pin']), GPIO.OUT)
    except:
      print('GPIO setup error for pin',switches[switch]['pin'])
# Can directly use the Boolean variabe onoff since True=1=GPIO.HIGH
    s.enterabs(stime,1,GPIO.output,argument=(int(switches[switch]['pin']),onoff));
#    print ('GPIO:',stime,switches[switch]['pin'],onoff);

def gpio_pulse(switch,onoff,stime):
# Set the pin to output (just to be sure...)
    try:
      GPIO.setup(int(switches[switch]['pin']), GPIO.OUT)
    except:
      print('GPIO setup error for pin',switches[switch]['pin'])
# Get the duration of the pulse
    if onoff:
      pulsetime=float(switches[switch]['on'])
    else:
      pulsetime=float(switches[switch]['off'])
# Chek for maximum pulse length, e.g. 10s (configured in switches.ini)
    if pulsetime>float(switches['DEFAULT']['max_pulse']):
      print ('The pulse duration of',pulsetime,'s is too long, setting to max=',max_pulse);
      pulsetime=float(switches['DEFAULT']['max_pulse']);
    s.enterabs(stime,1,GPIO.output,argument=(int(switches[switch]['pin']),1));
    s.enterabs(stime+pulsetime,1,GPIO.output,argument=(int(switches[switch]['pin']),0));

def dummy_switch(switch,onoff,stime):
    s.enterabs(stime,1,print,argument=('Dummy: ',onoff));

# Switch command options
# usage: switch_type[type]()
switch_type = {
  'rc'    : rc_switch,
  'gpio'  : gpio_switch,
  'pulse' : gpio_pulse,
  'dummy' : dummy_switch,
  }
 
switches = 0

#############################################################
# MAIN                                                      #
#############################################################
def main():

  GPIO.setmode(GPIO.BCM)
  GPIO.setwarnings(False)

  # Timezone offset
  tzoffset = datetime.today().hour-datetime.utcnow().hour

  # Read ini file for RC switch definition
  # keys: oncode, offcode, protocol, pulselength
  # example: switches['switchname']['oncode']
  global switches
  switches = configparser.ConfigParser()
  switches.sections()
  switches.read('/etc/caltimer/switches.ini')
  # Caldav url
  try:
    url = switches['DEFAULT']['caldav']
  except:
    print('Missing or incorrect ini file, please check /etc/caltimer/switches.ini')
    return
  # Scheduler intervall in Minuten
  interval = int(switches['DEFAULT']['interval'])
  # Maximum pulse length for GPIO pulses
  # max_pulse = float(switches['DEFAULT']['max_pulse'])

  # event options
  event_options = configparser.ConfigParser()


  client = caldav.DAVClient(url)
  try:
    principal = client.principal()
  except:
    e = sys.exc_info()[0]
    print('Error to access the web calendar:', e )
    return
  calendars = principal.calendars()
  if len(calendars) == 0:
    print ('No calender found at URL:',url)
    return
  
  if len(calendars) > 0:
    calendar = next((c for c in calendars if c.name == switches['DEFAULT']['calname']), None)
#    if type(calendar) ==  type(None)  :
    if calendar is  None:
      print('Calendar',switches['DEFAULT']['calname'],'not found.')
      print('Available calendars:')
      for calendar in calendars:
        print(calendar.name)
      return
    print ("Using calendar", calendar)

    dt = datetime.today()
    dt = dt + timedelta(minutes = interval - dt.minute % interval, seconds = -(dt.second % 60), microseconds = -(dt.microsecond % 1000000))
    dt_end = dt+timedelta(minutes=interval)

    print ("Events between: ", dt,"and",dt_end,"\n")

    results = calendar.date_search(
        dt - timedelta(hours=tzoffset), dt_end - timedelta(hours=tzoffset) )

    r_time_1 = 0
    r_time_2 = 0
    
    #print (len(results))
    if len(results)>0:
        # calculate sunrise and sunset
        ro = SunriseSunset(datetime.now(), latitude=float(switches['DEFAULT']['latitude']),
            longitude=float(switches['DEFAULT']['longitude']), localOffset=float(switches['DEFAULT']['local_offset']))
        rise_time, set_time = ro.calculate()
        print ("Sunrise",rise_time,", Sunset", set_time,"\n")
        if False: # print event list
          for event in results:
            event.load()
            e = event.instance.vevent
            print ("Start: " + e.dtstart.value.strftime("%H:%M") + ", Ende: " + e.dtend.value.strftime("%H:%M"))
            print ("Betreff:"+ e.summary.value + ", Ort:" + e.location.value);
            print ("Beschreibung:\n"+ e.description.value + "\n")
            if not e.location.value in switches:
              print ("Undefinerter RC-Switch",e.location.value)
            
    # schedule events
        print_time("Define scheduler at")
        global s
        s = sched.scheduler(time.time, time.sleep)
        for event in results:
            event.load()
            e = event.instance.vevent
            e_start = e.dtstart.value.timestamp()
            e_end = e.dtend.value.timestamp()
            # check if start/stop events are in current time interval
            s_start = e_start >= dt.timestamp() and e_start < dt_end.timestamp()
            s_end = e_end >= dt.timestamp() and e_end < dt_end.timestamp()

            try:
              event_options.read_string(e.description.value)
            except:
              print('Description undefined or incorrect')
              event_options.read_string('[DEFAULT]') # read dummy settings to clear data from previous event
# print event options
#            for each_section in event_options.sections():
#              for (each_key, each_val) in event_options.items(each_section):
#                print (each_key,':',each_val)

            r_time_1 = 0
            r_time_2 = 0
            # if there are random numbers define
            if event_options.has_section('random'):
              if event_options.has_option('random','all'):
                try:
                  r_time_1 = uniform(0,float(event_options['random']['all'])*60) # calculate random value from defined range
                  r_time_2 = r_time_1
                except:
                  print ("Random all is incorrect! Format is 'all : 999'")
              if event_options.has_option('random','start'):
                try:
                  r_time_1 = uniform(0,float(event_options['random']['start'])*60) # calculate random value from defined range
                except:
                  print ("Random start is incorrect! Format is 'start : 999'")
              if event_options.has_option('random','end'):
                try:
                  r_time_2 = uniform(0,float(event_options['random']['end'])*60) # calculate random value from defined range
                except:
                  print ("Random end is incorrect! Format is 'end : 999'")
            if event_options.has_section('sun'):
              if event_options.has_option('sun','start'):
                if event_options['sun']['start'] == "rise":
                  e_start=rise_time.timestamp()
                elif event_options['sun']['start'] == "set":
                  e_start=set_time.timestamp()
                else:
                  print('Sunrise start time option is incorrect, valid options are "rise" or "set"')
                try: # add start offset
                  e_start=e_start+(float(event_options['sun']['start_offset'])*60) # sunrise + offset
                except:
                  print ("Sunrise start offset format is incorrect! Format is 'start_offset : 999'")
            if event_options.has_section('sun'):
              if event_options.has_option('sun','end'):
                if event_options['sun']['end'] == "rise":
                  e_end=rise_time.timestamp()
                elif event_options['sun']['end'] == "set":
                  e_end=set_time.timestamp()
                else:
                  print('Sunrise end time option is incorrect, valid options are "rise" or "set"')
                try: # add end offset
                  e_end=e_end+(float(event_options['sun']['end_offset'])*60) # sunset + offset  
                except:
                  print ("Sunset end offset format is incorrect! Format is 'end_offset : 999'")

            if s_start:
                print (e.location.value,"einschlaten um",datetime.fromtimestamp(e_start).strftime('%H:%M:%S'),"{0:+.1f} min".format(r_time_1/60))
#                try:
                switch_type[switches[e.location.value]['type']](e.location.value,True,e_start+r_time_1)
#                except:
#                  print('Error:',e.summary.value,'at',e_start,'+',r_time_1,'!')
            if s_end:
                print (e.location.value, "ausschalten",datetime.fromtimestamp(e_end).strftime('%H:%M:%S'),"{0:+.1f} min".format(r_time_2/60))
                try:
                  switch_type[switches[e.location.value]['type']](e.location.value,False,e_end+r_time_2)
                except:
                  print('Error:',e.summary.value,print_time(e_end))
#        print (s.queue)
        print_time("Start scheduler at")
        print()
        s.run()
    else:
        print ("No switching events in this time interval.")
  

if __name__ == "__main__":
    main()


nanocul_relay.py
#!/usr/bin/python3

#===============================================================================
# This script receives RF messages from Kopp transmitters at 868 MHzs 
# and send them to other low cost RF 433 Mhz receivers
# 
# Matthias Homann
# 21.12.2017
#  
#===============================================================================

import sys
import subprocess
import logging
import argparse
import configparser
import serial
from time import sleep
import time

import RPi.GPIO as GPIO
from rpi_rf import RFDevice

GPIO.setwarnings(False)

# constant definitions
pulse_comag = 350
pulse_zap = 187
kopp_time = '00100'

# set initial logging to stderr, level INFO
logging.basicConfig(
    stream=sys.stderr,
    format='%(asctime)s - %(levelname)s - %(module)s - %(message)s',
    level=logging.INFO)

#+ 01234567890123456789
#~ kr07C2AD1A30CC0F0328
#~ ||  ||||  ||    ++-------- Transmitter Code 2
#~ ||  ||||  ++-------------- Keycode
#~ ||  ++++------------------ Transmitter Code 1
#~ ++------------------------ kr wird von der culfw bei Empfang einer Kopp Botschaft als 


#===============================================================================
# WZ Lampe an:  kt104B130300100N
# WZ Lampe aus: kt004B130300100N
# 
# Springbrunnen an:  kt1031090300100N
# Springbrunnen aus: kt0031090300100N
# Tase 2 an   : ktB031090300100N
# Taste 2 aus : ktA031090300100N
#===============================================================================

# Serial port config moved to ini file
#port = "/dev/ttyUSB.Nano"
#ser = serial.Serial(port, 38400, timeout=0)


def rf_switch(switch, onoff):
    if onoff:
        sendcode = config[switch]['oncode']
    else:
        sendcode = config[switch]['offcode']
    logging.info('<<< rf_switch send RF code %s at via %s',
                 sendcode,
                 config[switch]['rf_code'])
    if config[switch]['rf_code'] == "rf433":
        subprocess.call([config['DEFAULT']['rf433'], sendcode,
                         config[switch]['protocol'],
                         config[switch]['pulselength']],)
    elif config[switch]['rf_code'] == "rpi-rf":
        rfdevice.tx_code(int(sendcode), int(config[switch]['protocol']),
                         int(config[switch]['pulselength']))
    else:
        logging.error(
            'rf_switch undefined rf_code for switch %s, check ini file!',
            switch)


def rf_comag(switch, onoff):
    # Comag code calculation:
    # switch 0 = binary "01"
    # switch 1 = binary "00"
    # ON  = 10 = binary "0001"
    # OFF = 01 = binary "0100"
    #
    # Example:
    # Channel   Socket    ON/OFF
    # 0 1 0 0 0 0 0 1 1 0 10/01

    # Create binary code
    bincode = config[switch]['system'] + config[switch]['receiver']
    if onoff:
        bincode += '10'
    else:
        bincode += '01'
    logging.debug('*** Comag binary code = %s', bincode)
    # translate
    sendcode = 0
    for c in bincode:
        sendcode = sendcode << 2
        if c == "0":
            sendcode = sendcode | 1
    logging.debug('*** Comag sendcode = %s', '{:08b}'.format(sendcode))
    logging.info('<<< rf_comag send RF code %s via %s',
                 sendcode,
                 config[switch]['rf_code'])
    if config[switch]['rf_code'] == "rf433":
        subprocess.call([config['DEFAULT']['rf433'],
                         str(sendcode), "1", str(pulse_comag)],)
    elif config[switch]['rf_code'] == "rpi-rf":
        rfdevice.tx_code(int(sendcode), 1, pulse_comag)
    else:
        logging.error(
            'rf_comag undefined rf_code for switch %s, check ini file!',
            switch)


def rf_zap(switch, onoff):
    # ZAP/REV code calculation:_
    # tristate
    #   0 = binary "00"
    #   1 = binary "11"
    #   F = binary "01"
    # ON  = tri 01 = binary "0011"
    # OFF = tri 10 = binary "1100"
    #
    # Example:
    # ZAP-Code   Channel (F=open)  | Key 5..1          | On=01 Off=10
    # tri-state  0   0   F   F   F | 1   F   F   0   0 | 0   1
    # binary     00  00  01  01  01| 11  01  01  00  00| 00  11
    #            = 000001010111010100000011

    # The ZAP switch config provides:
    # 'channel'  : tri-state
    # 'zap_base' : tri-state base for the key part
    # 'key'      : decimal number of the receiver

    # Create binary code
    sendcode = 0
    for c in config[switch]['channel']:
        sendcode = sendcode << 2
        if c == '1':
            sendcode = sendcode | 3
        elif c == 'F':
            sendcode = sendcode | 1
    key_code = list(config[switch]['zap_base'])
    key_code[5-int(config[switch]['key'])] = "1"
    for c in key_code:
        sendcode = sendcode << 2
        if c[0] == '1':
            sendcode = sendcode | 3
        elif c[0] == 'F':
            sendcode = sendcode | 1
    sendcode = sendcode << 4
    if onoff:
        sendcode = sendcode | 3
    else:
        sendcode = sendcode | 12
    logging.debug('*** ZAP sendcode = %s', '{:08b}'.format(sendcode))
    logging.info('<<< rf_zap send RF code %s via %s',
                 sendcode,
                 config[switch]['rf_code'])
    if config[switch]['rf_code'] == "rf433":
        subprocess.call([config['DEFAULT']['rf433'],
                         str(sendcode), "1", str(pulse_zap)],)
    elif config[switch]['rf_code'] == "rpi-rf":
        rfdevice.tx_code(sendcode, 1, pulse_zap)
    else:
        logging.error(
            'rf_zap undefined rf_code for switch %s, check ini file!', switch)

def rf_kopp(switch, onoff, stime):
    # Kopp code example
    # 
    # kt004B130300100N
    # |||||||||||||||+-- Print output J/N
    # ||||||||||+++++--- Key pressed in ms
    # ||||++++++-------- Transmitter Code 1 + 2
    # ||++-------------- Key code on/off
    # ++---------------- kt = nanocul command for Kopp transmit

    sendcode='kt'
    if onoff:
        if config.has_option(switch, 'key_on'):
            sendcode += config[switch]['key_on']
        else:
            # calculate key_on from key_off by adding 0x10
            sendcode += format(int(config[switch]['key_off'],base=16)+16,'X')
    else:
        sendcode += config[switch]['key_off']
    sendcode += config[switch]['transmit_1'] + config[switch]['transmit_1'] + kopp_time + 'N'
    logging.info('<<< rf_kopp send code %s to nanocul', sendcode)
    x = ser.write((sencode+'\n').encode())
    
def configure_logging(log_arg, update, file):
    loglevel = {
        'CRITICAL': 50,
        'ERROR':    40,
        'WARNING':  30,
        'INFO':     20,
        'DEBUG':    10,
        'NOTSET':    0
    }
    # check if logfile is defined and can be opened for read
    print(config['LOGGING']['nanocul_log'])
    try:
        logfile = open(config['LOGGING']['nanocul_log'], 'r')
        logfile.close()
        log_exists = True
        #-------------------------------------- print ('File exists',log_exists)
    except:
        log_exists = False
    # check if logfile is defined and can be opened for write/append
    try:
        logfile = open(config['LOGGING']['nanocul_log'], 'a')
        logfile.close()
        # Remove all handlers associated with the root logger object.
        for handler in logging.root.handlers[:]:
            logging.root.removeHandler(handler)
        # Reconfigure logging again, this time with a file.
        logging.basicConfig(
            filename=config['LOGGING']['nanocul_log'],
            level=logging.INFO,
            format='%(asctime)s - %(levelname)s - %(module)s - %(message)s')
    except:
        if log_exists:
            temp_log = config['LOGGING']['nanocul_log'][:-4] \
                + time.strftime("_%y-%m-%d_%H-%M")+".log"
            logging.error(
                'Logfile is already in use by other process, '
                'using temp logfile: %s',
                temp_log)
            try:  # try to open new logfile write
                logfile = open(temp_log, 'w')
                logfile.close()
                # Remove all handlers associated with the root logger object.
                for handler in logging.root.handlers[:]:
                    logging.root.removeHandler(handler)
                # Reconfigure logging again, this time with a file.
                logging.basicConfig(
                    filename=temp_log, level=logging.INFO,
                    format='%(asctime)s - %(module)s -'
                    ' %(levelname)s : %(message)s')
            except:
                logging.error(
                    'No write access for temp logfile, '
                    'using sdterr for logging.')
        else:
            logging.error('No (correct) filename defined, '
                          'using sdterr for logging.')
    # log level set as command line parameter?
    if log_arg is not None:
        try:
            logging.info('Set loglevel: %s', log_arg)
            logging.getLogger().setLevel(loglevel[log_arg.upper()])
            if update:
                config.set('LOGGING','loglevel',log_arg.upper())
                update_ini(file)
        except:
            print ('ERROR: Incorrect loging level specified, '
                   'using log evel "ERROR"')
            logging.getLogger().setLevel(logging.ERROR)
    # if not, set logging level as defined in caltimer.ini
    else:
        logging.info('---------------------------------'
                     '---------------------------------')
        try:
            logging.info('Set loglevel: %s', config['LOGGING']['loglevel'])
            logging.getLogger().setLevel(
                loglevel[config['LOGGING']['loglevel']])
        except:
            logging.error('No loglevel defined, using ERROR')
            logging.getLogger().setLevel(logging.ERROR)

#############################################################
# MAIN                                                      #
#############################################################
def main():
    
    # Comamnd line arguments
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('-i', '--init',
                        help='specify path/filename of ini file to read from',
                        default='/etc/pyscript/nanocul.ini')
    parser.add_argument('-l', '--log',
                        help='set log level (overwrites ini file)\n'
                        '  Supported levels are: '
                        'CRITICAL, ERROR, WARNING, INFO, DEBUG')
    parser.add_argument('-u', '--update',
                        help='Update ini file (e.g. log level)',
                        action='store_true')
    args = parser.parse_args()
    
    # Read ini file for RC switch definition
    # keys: oncode, offcode, protocol, pulselength
    # example: config['switchname']['oncode']
    global config
    config = configparser.ConfigParser()
    #config.sections()
    config.read(args.init)
    if len(config) <= 1:
        print ("ERROR: The specified ini file doesn't exit!")
        return
    
    # set logfile destination and log level
    configure_logging(args.log, args.update, args.init)

    if config.has_option('DEFAULT', 'pulselength'):
        pulse_comag=int(config['DEFAULT']['pulselength'])
        logging.debug('Setting pulse_comag = %s',pulse_comag)
        
    if config.has_option('DEFAULT', 'zap_pulse'):
        pulse_zap=int(config['DEFAULT']['zap_pulse'])
        logging.debug('Setting pulse_zap = %s',pulse_zap)

    if config.has_option('DEFAULT', 'kopp_time'):
        kopp_time = config['DEFAULT']['kopp_time'].zfill(5)
        logging.debug('Setting kopp_time = %s',kopp_time)

    if config.has_option('DEFAULT', 'ser_port'):
        logging.debug('Create serial interface %s',config['DEFAULT']['ser_port'])
        global ser
        ser = serial.Serial(config['DEFAULT']['ser_port'], 38400, timeout=0)

    # Enable RF transmitter
    global rfdevice
    rfdevice = RFDevice(int(config['DEFAULT']['gpio']))
    rfdevice.enable_tx()
    
    relays=[]
    #receive_code=dict()
    on_code=dict()
    off_code=dict()
    for i in config.sections():
        if config.has_option(i, 'relay'):
            relays.append(i)
            off_code[(config[i]['transmit_1']+config[i]['transmit_2']+
                      config[i]['key_off']).encode()]=config[i]['relay'].split(',')
            if config.has_option(i, 'key_on'):
                on_code[(config[i]['transmit_1']+config[i]['transmit_2']+
                         config[i]['key_on']).encode()]=config[i]['relay'].split(',')
            else:
                # calculate key_on from key_off by adding 0x10
                key_on=format(int(config[i]['key_off'],base=16)+16,'X')
                on_code[(config[i]['transmit_1']+config[i]['transmit_2']+
                         key_on).encode()]=config[i]['relay'].split(',')
                print(key_on,config[i]['key_off'])
            #receive_code[i]=config[i]['transmit_1']+config[i]['transmit_2']
            #print(receive_code[i],on_code[i],off_code[i])
            logging.debug('Transmitter %s is relayed to %s',i,config[i]['relay'].split(','))
    logging.info('Relays found: %s',relays)
    logging.debug('ON:  %s', on_code)
    logging.debug('OFF: %s', off_code)
    
    logging.debug('Enable Kopp receive mode.')
    kr_enable=True
    while kr_enable: 
        x = ser.write(b'krS\n')
        sleep(0.2)
        data = ser.read(9999)
        if len(data) > 0:
            logging.debug('Received: %s',data)
            if data == b'krS-ReceiveStart\r\n':
                kr_enable=False
                logging.info('Kopp receive mode enabled!')

    logging.debug('Start to listen for receiver nanocul...')
    while True:
        data = ser.read(9999)
        if len(data) > 0:
            logging.debug('length: %s',len(data))
            #logging.debug('Got: %s', data.decode('utf-8')[0:-2]) # decode ascii or utf-8, strip last 2 char (cr lf)
            if len(data) == 22:
                transmitter = data[4:8] + data[16:18]
                keycode = data[10:12]
                rec_code=transmitter+keycode
                logging.debug('Transmitter: %s, Key %s', transmitter.decode(), keycode.decode())
                if rec_code in on_code:
                    logging.debug('ON: %s',on_code[rec_code])
                    for tx in on_code[rec_code]:
                        rf_zap(tx, True)
                        logging.info('Switch light %s ON',tx)
                if rec_code in off_code:
                    logging.debug('ON: %s',off_code[rec_code])
                    for tx in off_code[rec_code]:
                        rf_zap(tx, False)
                        logging.info('Switch light %s OFF',tx)
            else:
                logging.debug('Got: %s', data.decode()[0:-2]) # decode to string, strip last 2 char (cr lf)
    
        sleep(0.1) # check for new received code 10x per second
    
    logging.debug('Disable Kopp receive mode.')
    x = ser.write(b'krE\n')
    ser.close()

if __name__ == '__main__':
    main()

nancul-receive.py
#! /usr/bin/python3
import serial
from time import sleep

#+ 01234567890123456789
#~ kr07C2AD1A30CC0F0328
#~ ||  ||||  ||    ++-------- Transmitter Code 2
#~ ||  ||||  ++-------------- Keycode
#~ ||  ++++------------------ Transmitter Code 1
#~ ++------------------------ kr wird von der culfw bei Empfang einer Kopp Botschaft als 

#port = "/dev/ttyUSB.FTDI"
port = "/dev/ttyUSB.Nano"
ser = serial.Serial(port, 38400, timeout=0)

while True:
    data = ser.read(9999)
    if len(data) > 0:
        #print ('length:',len(data))
        #print ('Got:', data.decode('utf-8')[0:-2]) # decode ascii or utf-8, strip last 2 char (cr lf)
        if len(data) == 22:
            transmitter = data.decode('utf-8')[4:8] + data.decode('utf-8')[16:18]
            keycode = data.decode('utf-8')[10:12]
            print ('Transmitter:',transmitter,' Key:',keycode)
        else:
            print ('Got:', data.decode('utf-8')[0:-2]) # decode ascii or utf-8, strip last 2 char (cr lf)

    sleep(0.2)
    #print ('not blocked')

ser.close()

